### PR TITLE
Disable require.resolve for scopehoisting for bundled assets

### PR DESCRIPTION
**Pull request by mischnic**
# ↪️ Pull Request

- Disable `require.resolve` in scope hoisting for bundled assets
	- keep them (as they were) for excluded modules
	- as part of that, also call `addDependency` for `require.resolve` calls in the non-scope hoisting case

**Is there another usecase of `require.resolve` apart from `require(require.resolve(".."))`?**

#### Exhibit A
The existing test in
https://github.com/parcel-bundler/parcel/blob/23ae3e7d808936c18a3a461328399f5bbce01fc2/packages/core/integration-tests/test/integration/scope-hoisting/commonjs/require-resolve/a.js#L1-L3
doesn't make sense because the output is (what should I do with that hash, the corresponding module is already optimized away?):
```js
(function(){
output = "b4cf68639cd4b7a94bf4aa1dd2a041c3";
})();
```

#### Exhibit B

The only situation where `require.resolve` is useful is (modifying the test above):
```js
const foo =require('./b');
// ...
output = require(require.resolve('./b'));
```
however, that might be difficult to add to the existing scope hoisting flow and would above all would prevent some treeshaking.
Current output:
```js
(function(){
output = require("b4cf68639cd4b7a94bf4aa1dd2a041c3");
})();
```
(So this doesn't do at all what a user would expect, calling `require` on the result doesn't do the import)

#### Exhibit B'

Actually, (B) has another quirk in the current implementation:
```js
output = require(require.resolve('./b'));
```

actually throws (because `require.resolve` is a kind of weak import):
```js
Error: Got unexpected undefined
    at nullthrows (.../parcel/node_modules/nullthrows/nullthrows.js:7:15)
    at CallExpression (.../parcel/packages/shared/scope-hoisting/src/link.js:397:19)
```

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>packages/benchmarks/kitchen-sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 8.84s | +13.00ms |
| Cached | 8.53s | -17.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.ff215454.webp | 102.94kb | +0.00b | 208.00ms | -25.00ms 🚀 |
| dist/modern/parcel.76ee4591.webp | 102.94kb | +0.00b | 207.00ms | -26.00ms 🚀 |
| dist/legacy/index.js | 243.00b | +0.00b | 440.00ms | -61.00ms 🚀 |
| dist/modern/index.js | 215.00b | +0.00b | 435.00ms | -64.00ms 🚀 |
| dist/modern/index.css | 29.00b | +0.00b | 1.17s | -116.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.ff215454.webp | 102.94kb | +0.00b | 14.00ms | +1.00ms ⚠️ |
| dist/modern/parcel.76ee4591.webp | 102.94kb | +0.00b | 12.00ms | -1.00ms 🚀 |
| dist/modern/index.js | 215.00b | +0.00b | 13.00ms | -1.00ms 🚀 |
| dist/legacy/index.css | 29.00b | +0.00b | 12.00ms | +1.00ms ⚠️ |

</p></details>

<details><summary>packages/benchmarks/react-hn ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 14.18s | +571.00ms |
| Cached | 9.73s | +181.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/module/index.js | 34.02kb | +0.00b | 2.06s | +142.00ms ⚠️ |
| dist/main/index.js | 34.00kb | +0.00b | 2.06s | +142.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/module/index.js | 34.03kb | +0.00b | 10.00ms | +1.00ms ⚠️ |

</p></details>

<details><summary>packages/benchmarks/ak-editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.23m | +1.76s |
| Cached | 10.11s | -208.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 2.33mb | +430.00b ⚠️ | 1.83m | +979.00ms |
| dist/media-viewer.0f84e385.js | 86.78kb | +0.00b | 7.20s | +629.00ms ⚠️ |
| dist/card.424034bf.js | 76.35kb | +0.00b | 7.20s | +629.00ms ⚠️ |
| dist/popup.49173ea4.js | 44.40kb | +0.00b | 37.96s | -2.21s 🚀 |
| dist/popup.4eaf5454.js | 43.89kb | +0.00b | 41.00s | +3.77s ⚠️ |
| dist/card.f99e1597.js | 5.89kb | +0.00b | 15.60s | +829.00ms ⚠️ |
| dist/ResourcedEmojiComponent.2d80adbd.js | 2.25kb | +0.00b | 7.15s | +626.00ms ⚠️ |
| dist/card.9520f336.js | 1.80kb | +0.00b | 15.61s | +828.00ms ⚠️ |
| dist/card.aae6ebb9.js | 1.04kb | +0.00b | 15.60s | +821.00ms ⚠️ |
| dist/media-viewer-analytics-error-boundary.dd752a9e.js | 777.00b | +0.00b | 15.60s | +824.00ms ⚠️ |
| dist/esm.976debcc.js | 178.00b | +0.00b | 7.09s | +619.00ms ⚠️ |
| dist/dropzone.976debcc.js | 177.00b | +1.00b ⚠️ | 1.83m | +1.16m ⚠️ |
| dist/clipboard.976debcc.js | 177.00b | +1.00b ⚠️ | 41.47s | +288.00ms |
| dist/browser.976debcc.js | 177.00b | +1.00b ⚠️ | 1.25m | +34.04s ⚠️ |
| dist/editorView.976debcc.js | 177.00b | -1.00b 🚀 | 41.00s | -33.32s 🚀 |
| dist/media-card-analytics-error-boundary.976debcc.js | 176.00b | -5.00b 🚀 | 42.67s | -1.10m 🚀 |
| dist/16.976debcc.js | 174.00b | -5.00b 🚀 | 41.47s | -432.00ms |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 2.34mb | +430.00b ⚠️ | 422.00ms | -45.00ms 🚀 |
| dist/pdfRenderer.668ca09a.js | 1.14mb | +0.00b | 248.00ms | -14.00ms 🚀 |
| dist/smartMediaEditor.60646891.js | 601.42kb | +0.00b | 169.00ms | -19.00ms 🚀 |
| dist/Toolbar.b4bc0528.js | 105.39kb | +0.00b | 130.00ms | -9.00ms 🚀 |
| dist/EmojiPickerComponent.31eba2df.js | 105.35kb | +0.00b | 135.00ms | -10.00ms 🚀 |
| dist/ui.4644d263.js | 73.77kb | +0.00b | 132.00ms | -10.00ms 🚀 |
| dist/smartMediaEditor.a9b18e74.js | 67.43kb | +0.00b | 131.00ms | -10.00ms 🚀 |
| dist/component.10b390dc.js | 46.17kb | +0.00b | 151.00ms | -9.00ms 🚀 |
| dist/popup.49173ea4.js | 44.40kb | +0.00b | 153.00ms | -10.00ms 🚀 |
| dist/popup.0ccc1fc2.js | 40.68kb | +0.00b | 152.00ms | -10.00ms 🚀 |
| dist/popup.9050dea3.js | 37.83kb | +0.00b | 156.00ms | -9.00ms 🚀 |
| dist/DatePicker.3fed8988.js | 27.58kb | +0.00b | 124.00ms | -17.00ms 🚀 |
| dist/EmojiPickerComponent.7e355e17.js | 26.67kb | +0.00b | 124.00ms | -17.00ms 🚀 |
| dist/esm.f0029a4d.js | 25.71kb | +0.00b | 153.00ms | -8.00ms 🚀 |
| dist/js.5c3f5692.js | 16.54kb | +0.00b | 157.00ms | -9.00ms 🚀 |
| dist/workerHasher.d7e20c98.js | 11.56kb | +0.00b | 154.00ms | -9.00ms 🚀 |
| dist/EmojiTypeAheadComponent.1c09c30d.js | 10.44kb | +0.00b | 123.00ms | -17.00ms 🚀 |
| dist/card.f99e1597.js | 5.89kb | +0.00b | 157.00ms | -10.00ms 🚀 |
| dist/component.a26e593f.js | 5.62kb | +0.00b | 150.00ms | -9.00ms 🚀 |
| dist/index.css | 3.38kb | +0.00b | 102.00ms | -16.00ms 🚀 |
| dist/component.8fa44754.js | 3.38kb | +0.00b | 150.00ms | -9.00ms 🚀 |
| dist/png-chunks-extract.4dfdb705.js | 3.33kb | +0.00b | 156.00ms | -10.00ms 🚀 |
| dist/feedback.4dda4a59.js | 2.93kb | +0.00b | 125.00ms | -17.00ms 🚀 |
| dist/EmojiUploadComponent.0460cc59.js | 2.91kb | +0.00b | 123.00ms | -17.00ms 🚀 |
| dist/component.ab65829f.js | 2.29kb | +0.00b | 150.00ms | -9.00ms 🚀 |
| dist/ResourcedEmojiComponent.2d80adbd.js | 2.25kb | +0.00b | 161.00ms | -9.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.c660ebe7.js | 1.92kb | +0.00b | 152.00ms | -8.00ms 🚀 |
| dist/card.9520f336.js | 1.80kb | +0.00b | 158.00ms | -9.00ms 🚀 |
| dist/16.dcacbe5c.js | 1.69kb | +0.00b | 137.00ms | -16.00ms 🚀 |
| dist/date.cfe6153f.js | 1.63kb | +0.00b | 134.00ms | -15.00ms 🚀 |
| dist/16.de3f3184.js | 1.62kb | +0.00b | 139.00ms | -17.00ms 🚀 |
| dist/images.c60a97a9.js | 1.57kb | +0.00b | 133.00ms | -15.00ms 🚀 |
| dist/workerHasher.7f0e2d92.js | 1.48kb | +0.00b | 154.00ms | -10.00ms 🚀 |
| dist/workerHasher.30e330ec.js | 1.48kb | +0.00b | 123.00ms | -17.00ms 🚀 |
| dist/workerHasher.854262ce.js | 1.48kb | +0.00b | 122.00ms | -17.00ms 🚀 |
| dist/workerHasher.64914391.js | 1.48kb | +0.00b | 121.00ms | -16.00ms 🚀 |
| dist/status.e3ed4764.js | 1.35kb | +0.00b | 129.00ms | -16.00ms 🚀 |
| dist/list-number.dbe744df.js | 1.35kb | +0.00b | 132.00ms | -15.00ms 🚀 |
| dist/16.f54dac28.js | 1.34kb | +0.00b | 149.00ms | -10.00ms 🚀 |
| dist/16.8f0e8fe1.js | 1.29kb | +0.00b | 147.00ms | -10.00ms 🚀 |
| dist/16.43b62c01.js | 1.29kb | +0.00b | 138.00ms | -17.00ms 🚀 |
| dist/code.092e0137.js | 1.28kb | +0.00b | 135.00ms | -14.00ms 🚀 |
| dist/16.b17a4d30.js | 1.27kb | +0.00b | 136.00ms | -15.00ms 🚀 |
| dist/16.51bbc68e.js | 1.23kb | +0.00b | 136.00ms | -15.00ms 🚀 |
| dist/component.8351f228.js | 1.23kb | +0.00b | 151.00ms | -8.00ms 🚀 |
| dist/link.6fa27bb2.js | 1.20kb | +0.00b | 132.00ms | -15.00ms 🚀 |
| dist/heading6.3996fe48.js | 1.19kb | +0.00b | 125.00ms | -17.00ms 🚀 |
| dist/16.385d3ff5.js | 1.19kb | +0.00b | 137.00ms | -15.00ms 🚀 |
| dist/heading3.9fea5413.js | 1.18kb | +0.00b | 126.00ms | -17.00ms 🚀 |
| dist/16.b3ce4050.js | 1.16kb | +0.00b | 139.00ms | -17.00ms 🚀 |
| dist/16.e37bbe6b.js | 1.15kb | +0.00b | 137.00ms | -15.00ms 🚀 |
| dist/16.f7fe274a.js | 1.14kb | +0.00b | 135.00ms | -15.00ms 🚀 |
| dist/emoji.c4bc5355.js | 1.12kb | +0.00b | 133.00ms | -15.00ms 🚀 |
| dist/16.ac2ff83f.js | 1.10kb | +0.00b | 137.00ms | -16.00ms 🚀 |
| dist/16.4c652b29.js | 1.10kb | +0.00b | 138.00ms | -17.00ms 🚀 |
| dist/16.24165259.js | 1.09kb | +0.00b | 138.00ms | -18.00ms 🚀 |
| dist/16.74510c56.js | 1.09kb | +0.00b | 149.00ms | -10.00ms 🚀 |
| dist/16.97c5bdf9.js | 1.09kb | +0.00b | 137.00ms | -16.00ms 🚀 |
| dist/16.a79db876.js | 1.09kb | +0.00b | 148.00ms | -9.00ms 🚀 |
| dist/heading5.2f045997.js | 1.06kb | +0.00b | 126.00ms | -16.00ms 🚀 |
| dist/16.6f8ac62a.js | 1.05kb | +0.00b | 149.00ms | -9.00ms 🚀 |
| dist/card.aae6ebb9.js | 1.04kb | +0.00b | 158.00ms | -10.00ms 🚀 |
| dist/16.3e36da0b.js | 1.01kb | +0.00b | 136.00ms | -15.00ms 🚀 |
| dist/heading2.2194210a.js | 1.00kb | +0.00b | 127.00ms | -16.00ms 🚀 |
| dist/mention.9fdad24e.js | 981.00b | +0.00b | 130.00ms | -17.00ms 🚀 |
| dist/fallback.305c8864.js | 980.00b | +0.00b | 128.00ms | -16.00ms 🚀 |
| dist/heading4.63d1ca06.js | 976.00b | +0.00b | 126.00ms | -17.00ms 🚀 |
| dist/layout.6f66c2da.js | 962.00b | +0.00b | 133.00ms | -14.00ms 🚀 |
| dist/divider.160f3108.js | 940.00b | +0.00b | 134.00ms | -14.00ms 🚀 |
| dist/quote.88bfdca4.js | 936.00b | +0.00b | 129.00ms | -16.00ms 🚀 |
| dist/action.146450ba.js | 919.00b | +0.00b | 135.00ms | -15.00ms 🚀 |
| dist/decision.06254b93.js | 898.00b | +0.00b | 134.00ms | -15.00ms 🚀 |
| dist/panel-warning.59c8ef3f.js | 891.00b | +0.00b | 130.00ms | -16.00ms 🚀 |
| dist/list.4f623de3.js | 870.00b | +0.00b | 131.00ms | -15.00ms 🚀 |
| dist/heading1.65fb8482.js | 864.00b | +0.00b | 128.00ms | -16.00ms 🚀 |
| dist/panel-error.c2479698.js | 791.00b | +0.00b | 130.00ms | -17.00ms 🚀 |
| dist/panel.23cd7224.js | 788.00b | +0.00b | 130.00ms | -15.00ms 🚀 |
| dist/table.4ace5843.js | 780.00b | +0.00b | 128.00ms | -16.00ms 🚀 |
| dist/media-viewer-analytics-error-boundary.dd752a9e.js | 777.00b | +0.00b | 158.00ms | -10.00ms 🚀 |
| dist/panel-success.6a7b97d7.js | 728.00b | +0.00b | 129.00ms | -17.00ms 🚀 |
| dist/panel-note.5ff85bdc.js | 727.00b | +0.00b | 130.00ms | -16.00ms 🚀 |
| dist/simpleHasher.efc670cc.js | 446.00b | +0.00b | 154.00ms | -9.00ms 🚀 |
| dist/simpleHasher.f8edc084.js | 446.00b | +0.00b | 122.00ms | -18.00ms 🚀 |
| dist/simpleHasher.884d4eb6.js | 446.00b | +0.00b | 122.00ms | -17.00ms 🚀 |
| dist/simpleHasher.c5e45ebc.js | 446.00b | +0.00b | 121.00ms | -16.00ms 🚀 |
| dist/esm.976debcc.js | 176.00b | +2.00b ⚠️ | 160.00ms | -10.00ms 🚀 |
| dist/media-card-analytics-error-boundary.976debcc.js | 176.00b | +2.00b ⚠️ | 157.00ms | -9.00ms 🚀 |
| dist/dropzone.976debcc.js | 176.00b | +2.00b ⚠️ | 152.00ms | -9.00ms 🚀 |
| dist/clipboard.976debcc.js | 176.00b | +2.00b ⚠️ | 152.00ms | -8.00ms 🚀 |
| dist/browser.976debcc.js | 176.00b | +2.00b ⚠️ | 151.00ms | -9.00ms 🚀 |
| dist/16.976debcc.js | 176.00b | +2.00b ⚠️ | 135.00ms | -15.00ms 🚀 |
| dist/editorView.976debcc.js | 176.00b | +2.00b ⚠️ | 120.00ms | -16.00ms 🚀 |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/250321877810020865)**

File: packages/shared/scope-hoisting/src/link.js [hunk 1]
```diff
@@ -399,8 +400,18 @@ export function link({
             .getDependencies(mapped)
             .find(dep => dep.moduleSpecifier === source.value)
         );
-        let mod = nullthrows(bundleGraph.getDependencyResolution(dep));
-        path.replaceWith(t.valueToNode(mod.id));
+        if (!bundleGraph.getDependencyResolution(dep)) {
+          // was excluded from bundling
+          path.replaceWith(
+            REQUIRE_RESOLVE_CALL_TEMPLATE({ID: t.stringLiteral(source.value)})
```
Comment (devongovett) [inline packages/shared/scope-hoisting/src/link.js]: Should we only do this with CommonJS outputFormat? We should probably also check if the specifier starts with `.` (i.e. local file?). Probably wouldn't work once bundled...

File: packages/shared/scope-hoisting/src/link.js [hunk 2]
```diff
@@ -399,8 +400,18 @@ export function link({
...
+          );
+        } else {
+          throw new Error(
+            "`require.resolve` calls for local assets aren't supported with scope hoisting"
+          );
+          // let mod = nullthrows(bundleGraph.getDependencyResolution(dep));
+          // path.replaceWith(t.valueToNode(mod.id));
```
Comment (devongovett) [inline packages/shared/scope-hoisting/src/link.js]: Remove commented lines.

File: packages/shared/scope-hoisting/src/link.js (hunk 1 above)
Comment (mischnic) [reply packages/shared/scope-hoisting/src/link.js]: > Should we only do this with CommonJS outputFormat?

Done

> We should probably also check if the specifier starts with

A local file would have a dependency resolution, and the else clause always throws?

Comment (mischnic) [reply packages/shared/scope-hoisting/src/link.js]: > Should we only do this with CommonJS outputFormat?

Done

> We should probably also check if the specifier starts with

A local file would have a dependency resolution, and the else clause always throws?

**Pull request merged**